commit 877ee4535ceaf6f0dd05cc4c53a79dfc99a82ebc (refs/changes/85/4785/2)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2018-09-04T18:27:13-07:00 (1 year, 1 month ago)
    
    TRITON-749 CNS reaper marks things as reaped *before* reaping them

diff --git a/lib/reaper-stream.js b/lib/reaper-stream.js
index 8cf49ee..deb2341 100644
--- a/lib/reaper-stream.js
+++ b/lib/reaper-stream.js
@@ -178,6 +178,7 @@ ReaperFSM.prototype.state_checkLastVisited = function (S) {
 
 ReaperFSM.prototype.state_checkReaped = function (S) {
 	var self = this;
+	var log = self.log.child({uuid: self.vmuuid});
 	S.timeout(1000, function () {
 		self.lastError = new Error(
 		    'Timed out waiting for redis response');
@@ -198,8 +199,19 @@ ReaperFSM.prototype.state_checkReaped = function (S) {
 		 * forget that it existed now.
 		 */
 		if (val !== null) {
-			self.redis.del('vm:' + self.vmuuid);
-			S.gotoState('next');
+			self.redis.hget('vm:' + self.vmuuid, 'last_recs',
+			    S.callback(function (err2, val2) {
+				if (err2 || val2 === null || val2 === '{}') {
+					self.redis.del('vm:' + self.vmuuid);
+					S.gotoState('next');
+					return;
+				}
+
+				log.warn({ last_recs: val2 }, 'tried to ' +
+				    'delete reaped VM, but last_recs is not ' +
+				    'empty');
+				S.gotoState('fetchAndPush');
+			}));
 			return;
 		}
 
@@ -224,16 +236,16 @@ ReaperFSM.prototype.state_fetchAndPush = function (S) {
 			return;
 		}
 
-		if (obj.state === 'destroyed' || obj.destroyed ||
-		    obj.state === 'failed' || obj.state === 'incomplete') {
-			self.redis.hset('vm:' + self.vmuuid, 'reaped', 'yes');
-		}
-
 		if (self.stream.push(obj) === false) {
 			S.gotoState('sleep_full');
 			return;
 		}
 
+		if (obj.state === 'destroyed' || obj.destroyed ||
+		    obj.state === 'failed' || obj.state === 'incomplete') {
+			self.redis.hset('vm:' + self.vmuuid, 'reaped', 'yes');
+		}
+
 		S.gotoState('sleep');
 	}));
 };
